spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/otlp/OtlpAutoConfiguration.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2012-2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.boot.actuate.autoconfigure.tracing.otlp;
+
+import io.micrometer.tracing.otel.bridge.OtelTracer;
+import io.opentelemetry.api.OpenTelemetry;
+import io.opentelemetry.exporter.otlp.http.trace.OtlpHttpSpanExporter;
+import io.opentelemetry.exporter.otlp.trace.OtlpGrpcSpanExporter;
+import io.opentelemetry.sdk.trace.SdkTracerProvider;
+
+import org.springframework.boot.actuate.autoconfigure.tracing.ConditionalOnEnabledTracing;
+import org.springframework.boot.autoconfigure.AutoConfiguration;
+import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+/**
+ * {@link EnableAutoConfiguration Auto-configuration} for OTLP. Brave does not support
+ * OTLP, so we only configure it for OpenTelemetry. OTLP defines three transports that are
+ * supported: gRPC (/protobuf), HTTP/protobuf, HTTP/JSON. From these transports HTTP/JSON
+ * is not supported by the OTel Java SDK, and it seems there are no plans supporting it in
+ * the future, see: <a href=
+ * "https://github.com/open-telemetry/opentelemetry-java/issues/3651">opentelemetry-java#3651</a>.
+ * Because this class configures components from the OTel SDK, it can only support what
+ * the OTel SDK does: gRPC and HTTP/protobuf.
+ *
+ * @author Jonatan Ivanov
+ * @since 3.1.0
+ */
+@AutoConfiguration
+@ConditionalOnEnabledTracing
+@ConditionalOnClass({ OtelTracer.class, SdkTracerProvider.class, OpenTelemetry.class })
+@EnableConfigurationProperties(OtlpProperties.class)
+public class OtlpAutoConfiguration {
+
+	@Configuration(proxyBeanMethods = false)
+	@ConditionalOnClass(OtlpHttpSpanExporter.class)
+	@ConditionalOnProperty(prefix = "management.otlp.tracing.http", name = "enabled", matchIfMissing = true)
```

Comment:
Hey. We usually don't use the `enabled` properties to switch on/off features if there is another signal. In your case, the `OtlpHttpSpanExporter` on the classpath could be that feature.

Reply:
Both `OtlpHttpSpanExporter` and `OtlpGrpcSpanExporter` come from the same artifact (`io.opentelemetry:opentelemetry-exporter-otlp`).

Reply:
Ah, i didn't notice that. Then we have two options:

* leave the `enabled` property
* Remove the default from `endpoint` and use this as a trigger

Reply:
I wonder if a third option would be to use a single property that takes an enum:

```
management.otlp.tracing.something=http
management.otlp.tracing.something=grpc
```

Reply:
This would limit users to choose between http and grpc. Don't know if both are a valid use case, though.

Reply:
No, I don't either. If it's valid, it could be a `Set<NewEnum>`:

```
management.otlp.tracing.something=grpc,http
```

If it isn't valid. A `NewEnum` property would be preferable as, unlike multiple `.enabled` properties, it prevents you from configuring something that's invalid.

Reply:
FWIW, OpenTelemetry's own [autoconfigure module](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/autoconfigure/README.md#otlp-exporter-span-metric-and-log-exporters) exposes `otel.exporter.otlp.protocol` property with options `grpc` and `http/protobuf` which suggests the choice is usually exclusive. So IMO it makes sense to take the same approach in Spring Boot, and consider multiple exporters scenario as something that requires manual configuration.

Reply:
I like the enum set idea, it makes the choice more clear/explicit.

I'm not sure why OTel does this, the SDK supports multiple exporters, right now Boot supports it as well (we also have auto-configuration for zipkin), I think we should able to follow this patterns instead what OTel does (we are also doing this in case of Micrometer registries).

Reply:
I'm genuinely curious - how frequent is the multiple exporter scenario really?

Seems to me that offering a property that accepts a set of protocols makes it really easy to misconfigure things in a way that you end up sending the traces twice to the same collector, which you likely won't notice easily as the collector will deal with the duplicates.

Reply:
I'm not sure about OTel, in Micrometer (slightly different scenario though) we see people using multiple registries.

To me, using the set of protocols makes this simpler and harder to misconfigure (than enabled flags) since it's explicit what transfer(s) are selected.

But I think I see your point and if I strictly apply this to Micrometer, there we only have one transport for the OTLP registry. And also, since the OTel SDK does not support every transports (HTTP/JSON is in spec. but not supported by the SDK), the Collectors need to support all 3 options. This drives me towards providing one option for OTLP and if people needs a different one for some reason, or two, they are just a `@Bean` away.

Reply:
fyi: I updated the PR to provide one OTLP exporter.

Reply:
> using the set of protocols makes this simpler and harder to misconfigure (than enabled flags)

Just to avoid confusion, I wasn't advocating for the enabled flags proposal, but for enum based solution originally proposed by Andy in https://github.com/spring-projects/spring-boot/pull/34508#discussion_r1129226532 - so a property that takes a single protocol value, which results in an appropriate exporter being registered. I still see it as a good thing to align with what OTel does themselves.

In case someone wants multiple exporters, they can manually declare exporter beans causing the auto-configuration to back off. If find that approach quite analogous to what auto-configuration does in some other places.